src/librustc/middle/const_eval.rs

```diff
@@ -1104,10 +854,20 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
 
           let mut call_args = NodeMap();
           for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_hint = match ty_hint {
+                  ExprTypeChecked => ExprTypeChecked,
+                  UncheckedExprNoHint | UncheckedExprHint(_) => {
+                      if let Some(hint) = tcx.ast_ty_to_ty_cache.borrow().get(&arg.ty.id) {
+                          UncheckedExprHint(hint)
```

Comment:
Yea I messed up here:

```rust
#![feature(const_fn)]

const fn bla<T>(t: T) -> T {
    t
}

fn main() {
    let x: [String; bla(5)];
}
```


causes

> thread 'rustc' panicked at 'bad ty_hint: Some(T), LitInt(1, UnsuffixedIntLit(Plus))', src/librustc/middle/const_eval.rs:1147

>  The correct thing to do would be to get the type of callee and extract the fn types from there.

I'm not sure i understand. `callee` is taken from an [`ExprKind::Path`](http://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ExprKind.html#variant.Path), which most likely (as in my example) does not contain the generic args. Also, since typeck is running while this issue can occur, we might not have any additional information. All I can think of is to either trigger typeck for the expression or reimplement inference to infer the arguments from the return type and the explicit types in the args and path.

To continue doing what we did before, I can simply erase the type hint here. I can also improve a little on what we did before by checking if the type contains generics and only then erase the type hint. Then I'll add an issue about generics and const fn.